pkg/maps/metricsmap/metricsmap.go

```diff
@@ -187,4 +272,18 @@ func init() {
 		MaxEntries: MaxEntries,
 		Pinning:    ebpf.PinByName,
 	})
+
+	// Register metrics map collector only once
+	// This package is initialised when cilium cli commands are run (e.g. cilium service list)
+	// Since registering the collector takes approximately 1 second,
+	// it is executed in a separate goroutine to prevent introducing delays in cilium command execution
+	once.Do(func() {
+		go func() {
+			err := metrics.Register(newMetricsMapCollector())
+			if err != nil {
+				log.WithError(err).Error("Failed to register metrics map collector to Prometheus registry. " +
+					"cilium_datapath_drop/forward metrics will not be collected")
+			}
+		}()
+	})
```

Comment:
Hm, the fact that this initialization occurs when a Cilium command is run first seems undesirable. Ideally, the daemon initializes all the metrics it needs upfront. The Cilium commands (which are separate binaries from the daemon) should not have to run any code to initialize metrics. 

I think what we could do here is to export `newMetricsMapCollector()` and have to be called within the `pkg/metrics/cell.go` code which is imported by the Agent Cell in `daemon/cmd/cells.go`.

cc @dylandreimerink, does this sound reasonable?

Reply:
@dylandreimerink @christarazi Can you take a look [at latest commit](https://github.com/cilium/cilium/pull/27370/commits/54b1945191db724ac094c79e9785cf9f1b017d39). I have converted metricsmap to cell.Module which is invoked in `daemon/cmd/cells.go`. Now, only daemon will initialize metricsmap. I've tried to add it to `pkg/metrics/cell.go` as suggested, but than we have circular dependency issue
```
package github.com/cilium/cilium/api/v1/server
        imports github.com/cilium/cilium/pkg/metrics
        imports github.com/cilium/cilium/pkg/maps/metricsmap
        imports github.com/cilium/cilium/pkg/ebpf
        imports github.com/cilium/cilium/pkg/bpf
        imports github.com/cilium/cilium/pkg/controller
        imports github.com/cilium/cilium/pkg/metrics: import cycle not allowed
```
Furthermore, I've tried to add metricsmap collector to metricsmap module, like in Hive examples, with `cell.Metrics(newMetricsmapCollector)`, but I couldn't make it work that way. Even though all metricsmap metrics were registered, metricsmap Collector was not, so Collect method wasn't invoked on the scrape. I know that these metrics were registered because I've spawn separate goroutine that was incrementing counters periodically and I was able to scrape them. 
Is it possible to create custom Collector using metric types defined in `pkg/metrics/metric` package?

Reply:
The structure looks good. I think it's unfortunate that there's an import cycle, which suggests that further refactoring is needed, but let's not tackle that in this PR. It would be good to keep that as a note in a code comment on the metricsmap cell. 

I'll let Dylan get to the finer details here.